File: Java/Last-Stone-Weight/SwapSolution.java
class SwapSolution {
    private int removeLargest(List<Integer> stones) {
        //get the index of the largest element
        int indexOfLargest = stones.indexOf(Collections.max(stones));
        //get the value of the largest element
        int result = stones.get(indexOfLargest);
        //get the value of last element and replace it with the largest element
        stones.set(indexOfLargest, stones.get(stones.size() - 1));
        //remove the last element or largest element
        stones.remove(stones.size() - 1);
        return result;
    }

    public int lastStoneWeight(int[] stones) {
        List<Integer> stoneList = new ArrayList<>();
        //for each element in array stones
        for (int weight : stones) {
            //add to the arrayList stoneList
            stoneList.add(weight);
        }
        //loop as long as the size of the lise is no bigger than 1
        while (stoneList.size() > 1) {
            //remove the heaviest stone from stoneList and store it to stone1
            int stone1 = removeLargest(stoneList);
            //remove the heaviest stone from stoneList and store it to stone2
            int stone2 = removeLargest(stoneList);
            //if stone1 and stone2 do not have the same weight
            if (stone1 != stone2) {
                int newStone = stone1 - stone2;
                //add the new stone
                stoneList.add(newStone);
            }
        }
        //if stoneList is not empty, return first element in stoneList; otherwise, return 0
        return !stoneList.isEmpty() ? stoneList.remove(0) : 0;
    }
}

File: Python/Last-Stone-Weight/SwapSolution.py
class Solution(object):
    def lastStoneWeight(self, stones):
        """
        :type stones: List[int]
        :rtype: int
        """
        #helper function
        def remove_largest():
            #locate the index of the element with max value
            heaviest_1_idx = stones.index(max(stones))
            #swap the highest value element with the last element
            stones[heaviest_1_idx], stones[-1] = stones[-1], stones[heaviest_1_idx]
            #pop the element at the last index
            return stones.pop()
        
        #special case handler
        #list is empty
        if not stones:
            return 0
        #list has only 1 element
        elif len(stones) == 1:
            return stones[0]
        else:
            #loop as long as list has no more than 1 element
            while len(stones) > 1:
                #pop the heaviest element in the list
                heaviest_1 = remove_largest()
                #pop the heaviest element in the list
                heaviest_2 = remove_largest()
                #if the weights of two heaviest elements are not the same
                if heaviest_1 - heaviest_2 != 0:
                    #append the new element into the list after the different weight is subtract
                    stones.append(heaviest_1 - heaviest_2)
            #return the only element in the list if the list is not empty; otherwise, return 0
            return stones[0] if stones else 0
        
